base/inference.jl

```diff
@@ -52,20 +52,39 @@ const NF = NotFound()
 typealias LineNum Int
 typealias VarTable Array{Any,1}
 
+# The type of a variable load is either a value or an UndefVarError
 type VarState
     typ
     undef::Bool
     VarState(typ::ANY, undef::Bool) = new(typ, undef)
 end
 
-immutable Const
+# The type of a value might be constant
+type Const
```

Comment:
we may frequently move these objects between the heap and stack. inference also attempts to aggressively use egal pointer comparisons to check whether something has changed before doing a more expensive test on its type and value, so it can be very sensitive to the box getting copied.